### PR TITLE
--kbuildout option: support extract kernel head files when building Linux kernel out-of-tree

### PR DESCRIPTION
**Pull request by iotcg**
KBUILD_OUT is the variable that is used in Linux build system when building kernel source out-of-tree.
Now, "--kbuildout" option is used in adeb build system to extract the kernel head files when building kernel out-of-tree.

### CODE REVIEW

File: BCC.md [hunk 1]
```diff
@@ -25,6 +25,8 @@ run the following command pointing adeb to the kernel sources which will
 have it extract headers from there and push them to the device.
 ```
 adeb prepare --build --bcc --kernelsrc /path/to/kernel-source/
+OR(If you build kernel out-of-tree):
+adeb prepare --build --bcc --kbuildout /path/to/kernel-kbuild_out-dir/
```
Comment (joelagnel) [inline BCC.md]: Don't you need to point to both the in-tree and out-of-tree directories in this command?

File: androdeb [hunk 2]
```diff
@@ -268,6 +285,7 @@ if [ ! -z "$KERNELSRC" ]; then
 	tar -xvf $TDIR_ABS/kh.tgz -C $OUT_TMP/kernel-headers/ > /dev/null
 fi
 
+
```
Comment (joelagnel) [inline androdeb]: why this newline?

File: androdeb [hunk 3]
```diff
@@ -66,6 +67,12 @@ case $key in
 esac
 done
 
+if [ ! -z "$KERNELSRC" -a ! -z "KBUILDOUT" ];
```
Comment (joelagnel) [inline androdeb]: I prefer if its called "KERNELOUT" to be consistent with "KERNELSRC"

File: androdeb [hunk 4]
```diff
@@ -66,6 +67,12 @@ case $key in
...
+then
+   c_warning "--kernelsrc option will set --kbuildout option ineffective"
```
Comment (joelagnel) [inline androdeb]: why? I think you need both the kernelsrc and kernelout paths. kernelsrc also has headers that need to be extracted. Overall your patch seems to be incomplete.

Comment (iotcg) [reply androdeb]: When building kernel out of tree, "source" directory will be linked to the original kernel source directory.
kernel_source/Makefile
  489 ifneq ($(KBUILD_SRC),)
  490     $(Q)ln -fsn $(srctree) source
  491     $(Q)$(CONFIG_SHELL) $(srctree)/scripts/mkmakefile \
  492         $(srctree) $(objtree) $(VERSION) $(PATCHLEVEL)
  493 endif

File: androdeb (hunk 3 above)
Comment (iotcg) [reply androdeb]: I'll change it to be "KERNELOUT" as you suggest.
What's about the '--kbuildout'? Do we also need change it to be "--kernelout"?

File: BCC.md (hunk 1 above)
Comment (iotcg) [reply BCC.md]: No.
In my script, kbuildbout parameter is ignored after giving kernesrc parameter.
--kernelsrc will be used in the cases that kernel in built in-tree.
--kbuildout will used in the cases that kernel is build out-of-tree. When building kernel out-of-tree, "source" directory is linked to the original kernel source directory. The script will deal with this case.

File: androdeb (hunk 2 above)
Comment (iotcg) [reply androdeb]: Do you mean below line?
        268   285       tar -xvf $TDIR_ABS/kh.tgz -C $OUT_TMP/kernel-headers/ > /dev/n
If yes ==> This isn't new line. It used in the base code repository.
If no ==> Please tell me which line directly.

**Comment by iotcg** [comment]
@joelagnel 
I'll follow your idea to use KERNELOUT to be consistend with "KERNELSRC"
Before doing this, do you have some other comments to change the patches? I've replied you comment one by one. Let me know whether there are other changes to be done.

**Comment by iotcg** [comment]
@joelagnel Do you have some other comments to change the patches besides using "KERNELOUT" instead of "KERNELSRC"?

**Pull request closed without merging**